lib/vanagon/component/dockerfile.rb

```diff
@@ -0,0 +1,188 @@
+require 'vanagon/component'
+require 'vanagon/utilities/shell_utilities'
+
+class Vanagon
+  class Component
+    # Vanagon::Component::Dockerfile generates the Dockerfile contents for
+    # a given component.
+    #
+    # TODO: This class + the corresponding Component::Rules class have a similar
+    # structure, specifically in how they calculate the required commands that will
+    # need to run for a given step. That calculation logic should be factored out into
+    # a common module. Note that it is for precisely this reason that the individual
+    # "step" methods are essentially copy-pastes of the corresponding "rule" methods
+    # in Vanagon::Component::Rules.
+    class Dockerfile
+      include Vanagon::Utilities::ShellUtilities
+
+      # Create methods that generate the Dockerfile contents for a specific
+      # step when building the component.
+      #
+      # This method cuts some of the boilerplate command munging + Dockerfile
+      # generation for each step.
+      #
+      # The block should return an array of commands that will be run for a given
+      # step.
+      #
+      # @param step_name [String] The step's name
+      def self.step(step_name, &block)
+        define_method("#{step_name}_step") do
+          commands = instance_exec(&block)
+          commands.compact!
+          commands.reject!(&:empty?)
+          commands.flatten!
+          commands = andand_multiline(commands)
+
+          next "" if commands.empty?
+
+          <<-DOCKERFILE
+  # #{component.name}-#{step_name}
+  RUN #{commands}
+          DOCKERFILE
+        end
+      end
+
+      attr_accessor :component
+      attr_accessor :project
+      attr_accessor :platform
+
+      # @param component [Vanagon::Component] The component to generate the Dockerfile for.
+      # @param project [Vanagon::Project] The project associated with the component.
+      # @param platform [Vanagon::Platform] The platform where this component will be built.
+      def initialize(component, project, platform)
+        @component = component
+        @project = project
+        @platform = platform
+      end
+
+      # This override of andand_multiline is purely for aesthetic purposes, i.e. it makes
+      # the generated Dockerfile more readable.
+      def andand_multiline(*commands)
+        cmdjoin(commands, " && \\\n    ")
```

Comment:
I have this separator so that something like
```
RUN command1 && \
         command2
```

can be printed, but for some reason the newline is only indented by a single space, so
```
RUN command1 && \
   command2
```

is printed instead. I'll look into this later, but wanted to point that out here in case anyone knew why/how to fix it.